internal/pkg/peer/blocksprovider/bft_deliverer.go

```diff
@@ -85,76 +123,123 @@ func (d *BFTDeliverer) DeliverBlocks() {
+		d.censorshipMonitor = d.CensorshipDetectorFactory.Create(
+			d.ChannelID, d.BlockVerifier, d.requester, d, d.fetchSources, d.fetchSourceIndex, timeoutConfig)
+		go d.censorshipMonitor.Monitor()
 
 		// wait for fetch  errors, censorship suspicions, or a stop signal.
 		select {
 		case <-d.DoneC:
 			break FetchAndMonitorLoop
 		case errFetch := <-d.fetchErrorsC:
-			if errFetch != nil {
-				switch errFetch.(type) {
-				case *errStopping:
-					// nothing to do
-				case *errRefreshEndpoint:
-					// get new endpoints and reassign fetcher and monitor
-					d.refreshSources()
-					d.resetFetchFailureCounter()
-				default:
-					d.fetchSourceIndex = (d.fetchSourceIndex + 1) % len(d.fetchSources)
-					d.incFetchFailureCounter()
-				}
-				// TODO can it be nil?
+			d.Logger.Debugf("Error received from fetchErrorsC channel: %s", errFetch)
+
+			switch errFetch.(type) {
+			case *ErrStopping:
+				// nothing to do
+				break FetchAndMonitorLoop
+			case *errRefreshEndpoint:
+				// get new endpoints and reassign fetcher and monitor
+				d.refreshSources()
+				d.resetFetchFailureCounter()
+			case *ErrFatal:
+				d.Logger.Errorf("Failure in FetchBlocks, something is critically wrong: %s", errFetch)
+				break FetchAndMonitorLoop
+			default:
+				d.fetchSourceIndex = (d.fetchSourceIndex + 1) % len(d.fetchSources)
+				d.incFetchFailureCounter()
 			}
-		case errMonitor := <-d.monitorErrorsC:
-			// TODO until we implement the censorship monitor this nil channel is blocked
-			if errMonitor != nil {
-				d.Logger.Warningf("Censorship suspicion: %s", err)
-				// TODO close the block receiver, increment the index
-				// TODO
+		case errMonitor := <-d.censorshipMonitor.ErrorsChannel():
+			d.Logger.Debugf("Error received from censorshipMonitor.ErrorsChannel: %s", errMonitor)
+
+			switch errMonitor.(type) {
+			case *ErrStopping:
+				// nothing to do
+				break FetchAndMonitorLoop
+			case *ErrCensorship:
+				d.Logger.Warningf("Censorship suspicion: %s", errMonitor)
+				d.mutex.Lock()
+				d.blockReceiver.Stop()
+				d.mutex.Unlock()
+				d.fetchSourceIndex = (d.fetchSourceIndex + 1) % len(d.fetchSources)
+				d.incFetchFailureCounter()
+			case *ErrFatal:
+				d.Logger.Errorf("Failure in CensorshipMonitor, something is critically wrong: %s", errMonitor)
+				break FetchAndMonitorLoop
+			default:
+				d.Logger.Errorf("Unexpected error from CensorshipMonitor, something is critically wrong: %s", errMonitor)
+				break FetchAndMonitorLoop
 			}
-			// TODO can it be nil?
 		}
+
+		d.censorshipMonitor.Stop()
 	}
 
 	// clean up everything because we are closing
 	d.mutex.Lock()
 	defer d.mutex.Unlock()
 	d.blockReceiver.Stop()
-	// TODO stop the monitor
+	if d.censorshipMonitor != nil {
+		d.censorshipMonitor.Stop()
+	}
 }
 
 func (d *BFTDeliverer) refreshSources() {
 	// select an initial source randomly
-	d.fetchSources = shuffle(d.Orderers.Endpoints())
+	d.fetchSources = d.Orderers.ShuffledEndpoints()
```

Comment:
why do we shuffle each time we refresh an endpoint?

Reply:
When even a single endpoint is changed, all the endpoints are refreshed, i.e. respective channel is closed. In that case we have a new array of sources, possibly of different length. This happens following a config TX only, and the first time the bft deliverer is started, that is - it happens very rarely. Following a config TX the raw source slice will be the same in all peers. Therefore, in order to avoid all peers selecting the same orderer as their block source, we have to shuffle the raw sources slice and select from the shuffled slice.